File: Baekjoon Online Judge/easy/24060.java
import java.io.*;
import java.util.*;

public class Main {
    static int cnt = 0, K, res = -1;
    public static void main(String[] args) throws IOException {
        // Init
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // Input
        K = Integer.parseInt(br.readLine().split(" ")[1]);
        int[] A = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        // Solve
        mergeSort(A, 0, A.length - 1);

        // Output
        System.out.println(res);
    }
    private static void mergeSort(int[] arr, int left, int right) {
        if (left >= right) return;
        int mid = (left + right) >> 1;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr,
                Arrays.copyOfRange(arr, left, mid + 1),
                Arrays.copyOfRange(arr, mid + 1, right + 1),
                left);
    }
    private static void merge(int[] arr, int[] left, int[] right, int k) {
        int l = 0, r = 0;
        while (l < left.length && r < right.length) {
            arr[k++] = (left[l] <= right[r]) ? left[l++] : right[r++];
            if (++cnt == K) res = arr[k - 1];
        }
        while (l < left.length) {
            arr[k++] = left[l++];
            if (++cnt == K) res = arr[k - 1];
        }
        while (r < right.length) {
            arr[k++] = right[r++];
            if (++cnt == K) res = arr[k - 1];
        }
    }
}

File: baekjoon-online-judge/silver/24060.py
from sys import stdin

input = stdin.readline

def check_cnt():
    global cnt
    cnt += 1
    return cnt == K


def merge_sort(left, right):
    if left >= right:
        return
    mid = (left + right) // 2
    merge_sort(left, mid)
    merge_sort(mid + 1, right)
    merge(left, mid, right)


def merge(left, mid, right):
    global res
    temp = []
    l, r = left, mid + 1
    while l <= mid and r <= right:
        if arr[l] <= arr[r]:
            temp.append(arr[l])
            l += 1
        else:
            temp.append(arr[r])
            r += 1

    while l <= mid:
        temp.append(arr[l])
        l += 1

    while r <= right:
        temp.append(arr[r])
        r += 1

    # Update original array with sorted values
    for i in range(left, right + 1):
        arr[i] = temp[i - left]
        if check_cnt(): res = arr[i]

if __name__ == '__main__':
    # Input
    _, K = map(int, input().split())
    arr = list(map(int, input().split()))

    # Solve
    cnt, res = 0, -1
    merge_sort( 0, len(arr) - 1)

    # Output
    print(res)
